misc/bazel/cargo-gazelle/src/context.rs

```diff
@@ -0,0 +1,190 @@
+    /// What protobuf files, if any, this build script generates bindings for.
+    pub generated_protos: Vec<String>,
+    /// Relative paths of dependencies for the protobuf files we generate.
+    pub proto_dependencies: BTreeSet<Utf8PathBuf>,
+}
+
+impl BuildScriptContext {
+    pub fn generate(
+        config: &Config,
+        build_script_path: impl AsRef<Path>,
+    ) -> Result<BuildScriptContext, anyhow::Error> {
+        let build_script_path = build_script_path.as_ref();
+        tracing::debug!(?build_script_path, "generating build script context");
+
+        let mut context = BuildScriptContext::default();
+
+        let mut file = File::open(build_script_path)?;
+        let mut content = String::new();
+        file.read_to_string(&mut content)?;
+
+        // Parse the AST of the build script to find any invocations of protobuf generation and the
+        // files we're attempting to generate.
+        let ast = syn::parse_file(&content)?;
+        let proto_files: Vec<String> = ast
+            .items
+            .iter()
+            .filter_map(|item| match item {
+                syn::Item::Fn(func) => Some(func),
+                _ => None,
+            })
+            .flat_map(|func| {
+                func.block.stmts.iter().filter_map(|stmt| match stmt {
+                    syn::Stmt::Expr(expr) => Some(expr),
+                    _ => None,
+                })
+            })
+            .filter(|expr| find_proto_build(config, expr))
+            .map(|expr| find_proto_files(config, expr.to_token_stream()))
+            .flat_map(|files| files.into_iter())
+            // Remove the first component of the path.
+            .map(|s| {
+                tracing::debug!(path = s, "protobuf path");
+                let mut path = Utf8PathBuf::new();
+                for c in camino::Utf8PathBuf::from(s).components().skip(1) {
+                    path.push(c);
+                }
+                path.to_string()
+            })
+            .collect();
+
+        // Parse the protobuf files we just found to determine dependencies.
+        let crate_root_path = build_script_path
+            .parent()
+            .ok_or_else(|| anyhow!("build script at the root of the filesystem?"))?;
+        let proto_dependencies: BTreeSet<_> = proto_files
+            .iter()
+            .map(|sub_path| {
+                tracing::debug!(?sub_path, "protobuf dependency");
+                let full_path = crate_root_path.join(sub_path);
+                parse_proto_dependencies(&full_path).context("parsing build script proto")
+            })
+            .collect::<Result<Vec<_>, _>>()?
+            .into_iter()
+            .flat_map(|paths| paths.into_iter())
+            .collect();
+
+        context.generated_protos = proto_files;
+        context.proto_dependencies = proto_dependencies;
+
+        Ok(context)
+    }
+}
+
+// TODO(parkmycar): There is almost definitely a better way to do this AST parsing.
```

Comment:
In the long run: recommend simplifying the problem. The build.rs scripts are under our control. Instead of having each build.rs script imperatively compile protobufs, we could declare the protobufs to compile in the Cargo.toml in a custom metadata section, and then teach both build.rs and cargo-gazell to read the protobufs to compile from the Cargo.toml.

Reply:
This makes sense! To make the addition of Bazel as seamless as possible I was hoping to avoid any changes like this, but I think defining the protos in the `Cargo.toml` is still easy enough and makes the build process more robust 👍